### PR TITLE
Adds htaccess file viewer to tools

### PR DESCRIPTION
**Pull request by jocastaneda**
As the title says, it adds the `.htaccess` file viewer to the tools tab if the file exists and has any custom or additional text. It also provides the ability to copy the text just in case a user needs to copy/paste that.

fixes: https://github.com/WordPress/healtch-check/issues/178

### CODE REVIEW

File: src/includes/class-health-check-htaccess-viewer.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	 * @return array Associative array.
+	 */
+	public static function tools_tab( $tabs ) {
+		// check if it is a file.
+		if ( ! is_file( ABSPATH . '.htaccess' ) ) {
+			return $tabs;
+		}
+
+		// It does so let's get the contents
+		$htaccess_file = file_get_contents( ABSPATH . '/.htaccess' );
```
Comment (Clorith) [inline src/includes/class-health-check-htaccess-viewer.php]: I would add a check that we can actually read the file contents, as in some scenarios, it is "locked" from the web user.

File: src/includes/class-health-check-htaccess-viewer.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+		// Remove WordPress rules.
+		$filtered_htaccess_content = trim( preg_replace( '/\# BEGIN WordPress[\s\S]+?# END WordPress/si', '', $htaccess_file ) );
+
+		// Return since it is only core added rules.
+		if ( '' === $filtered_htaccess_content ) {
+			return $tabs;
```
Comment (Clorith) [inline src/includes/class-health-check-htaccess-viewer.php]: I'm wondering if we should allow the displaying of the content, even if it's only got core content, avoiding conditional displaying as often as possible for a clearer user experience?

**Review by Clorith — COMMENTED** [review]
I like it, just a few discussion points I've marked that I'd like to hear your thoughts on before moving on here.

File: src/includes/class-health-check-htaccess-viewer.php (hunk 1 above)
Comment (jocastaneda) [reply src/includes/class-health-check-htaccess-viewer.php]: true, will look into this either today or tomorrow.

File: src/includes/class-health-check-htaccess-viewer.php (hunk 2 above)
Comment (jocastaneda) [reply src/includes/class-health-check-htaccess-viewer.php]: So, display it regardless?

I'm fine with that. We'd need to add a message stating something for it too.

**Comment by jocastaneda** [comment]
@Clorith I think when there is an `.htaccess` file we will need to add at least something stating hey, there are core added and if there are more, let the user know as well.

**Comment by Clorith** [comment]
A short explanation of what the htaccess file does will probably suffice I'm thinking, since there's a "WordPress" wrapper around the core rules them selves, I wouldn't delve too deeply into semantics, as that falls a bit out of scope in my opinion.

**Comment by Clorith** [comment]
Would you please change the branch to `develop` before this can be implemented, to suit the plugins new development flow.

**Comment by jocastaneda** [comment]
Changed it. Will finally get a chance to work on this tonight! 😍

**Review by Clorith — CHANGES_REQUESTED** [review]

File: src/includes/class-health-check-debug-data.php [hunk 3]
```diff
@@ -461,9 +461,16 @@ static function debug_data( $locale = null ) {
...
+			$tools_tab = add_query_arg( array(
+				'page' => 'health-check',
+				'tab'  => 'tools',
+			), admin_url( 'index.php' ) );
+
 			$info['wp-server']['fields'][] = array(
 				'label' => __( 'htaccess rules', 'health-check' ),
-				'value' => ( ! empty( $filtered_htaccess_content ) ? __( 'Custom rules have been added to your htaccess file', 'health-check' ) : __( 'Your htaccess file only contains core WordPress features', 'health-check' ) ),
+				/* translators: admin url */
+				'value' => ! empty( $filtered_htaccess_content ) ? sprintf( __( 'Custom rules have been added to your htaccess file, you can view the contents <a href="%s">here</a>', 'health-check' ), esc_url( $tools_tab ) ) : __( 'Your htaccess file only contains core WordPress features', 'health-check' ),
```
[the remark this answers is not in the document]
Comment (jocastaneda) [reply src/includes/class-health-check-debug-data.php]: Addressed in lates commit

File: src/includes/class-health-check-htaccess-viewer.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	public static function tools_tab( $tabs ) {
+		// check if it is a file and readable.
+		if ( ! is_readable( ABSPATH . '.htaccess' ) ) {
+			$tabs['htaccess-viewer'] = array(
+				'label'   => esc_html__( 'Contents of .htaccess', 'health-check' ),
+				'content' => sprintf( '<div><p>%s</p></div>', esc_html__( 'It appears there is no .htaccess file or is not a readable file.', 'health-check' ) ),
+			);
+		} else {
+			// it is a file.
+			$htaccess_file = file_get_contents( ABSPATH . '/.htaccess' );
```
[the remark this answers is not in the document]
Comment (jocastaneda) [reply src/includes/class-health-check-htaccess-viewer.php]: Addressed in latest commit

File: src/includes/class-health-check-htaccess-viewer.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+			// Remove WordPress rules.
+			$filtered_htaccess_content = trim( preg_replace( '/\# BEGIN WordPress[\s\S]+?# END WordPress/si', '', $htaccess_file ) );
```
[the remark this answers is not in the document]
Comment (jocastaneda) [reply src/includes/class-health-check-htaccess-viewer.php]: Yes. Can't recall why it was there to begin. It has been removed in the latest commit as well

File: src/includes/class-health-check-htaccess-viewer.php [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+			ob_start(); ?>
+			<div>
+				<p>
+					<button type="button" class="button button-primary health-check-copy-field"><?php esc_html_e( 'Mark field for copying', 'health-check' ); ?></button>
+				</p>
+				<textarea id="htaccess-contents" class="widefat" rows="15"><?php echo htmlspecialchars( file_get_contents( ABSPATH . '.htaccess' ) ); ?></textarea>
```
[the remark this answers is not in the document]
Comment (jocastaneda) [reply src/includes/class-health-check-htaccess-viewer.php]: In the latest it uses `esc_html`. I'm not super picky about it

**Pull request closed without merging**